### PR TITLE
Added the ability to emit file ASTs as JSON

### PR DESCRIPTION
**Pull request by derekxu16**
Implemented and tested: variable declarations, interface declarations, class declarations, type literals, keyword types, module declarations

Began work on: type aliases, type references, imports/exports, function declarations

### CODE REVIEW

**Comment by terrylucas** [comment]
Can you attach the JSON payload generated?  How log does it take to generate the JSON file?

File: lib/json/named_declarations/parameter_declaration.ts [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+  private optional = false;
+  private destructured = false;
+  private initializer?: string;
+
+  constructor(node: tsParameterDeclaration) {
+    super(node, ConvertedSyntaxKind.Parameter);
+
+    this.type = convertTypeNode(node.type);
+    if (node.questionToken) {
+      // TODO(derekx): Do we want to mark parameters with default values as optional?
```
Comment (srujzs) [inline lib/json/named_declarations/parameter_declaration.ts]: All parameters with default values are inherently optional, right? I can see the argument that it should only be marked optional if it's explicitly optional, but I don't see an instance in the conversion to Dart where that separation makes a difference.

File: test/json/declarations/class_test.ts [hunk 2]
```diff
@@ -0,0 +1,490 @@
...
+            {kind: ConvertedSyntaxKind.TypeParameter, name: 'T1'},
+            {kind: ConvertedSyntaxKind.TypeParameter, name: 'T2'}
+          ],
+          members: []
+        }]
+      }));
+    });
+
+    it('should handle type parameters with constraints', () => {
+      expectTranslateJSON('declare class X<T extends string> {}').to.equal(prettyStringify({
```
Comment (srujzs) [inline test/json/declarations/class_test.ts]: Can the script handle nested generics e.g. Foo<Bar extends Baz<Boo extends ...>>?

Comment (srujzs) [reply test/json/declarations/class_test.ts]: Or I guess a better question: does it need to?

Comment (derekxu16) [reply test/json/declarations/class_test.ts]: It won't accept that syntax, but you can do something similar like
`interface X<U extends number, T extends Promise<U>> {}`

I'll add a test for this

File: lib/json/named_declarations/parameter_declaration.ts (hunk 1 above)
Comment (srujzs) [reply lib/json/named_declarations/parameter_declaration.ts]: Seems like you went with the option to not mark them optional (which is fine, I'm just commenting to resolve this question).

**Pull request merged**